[C++20][Modules][Serialization] Delay marking pending incomplete decl chains until the end of `finishPendingActions`.

mpark opened the pull request:
# Description

There is a defaulted constructor `_Hashtable_alloc() = default;` which ends up in [`CodeGenFunction::GenerateCode`](https://github.com/llvm/llvm-project/blob/48bf0a9457fd60d0872d9b9b4804a95c833a72e1/clang/lib/CodeGen/CodeGenFunction.cpp#L1448) and eventually calls [`FunctionProtoType::canThrow`](https://github.com/llvm/llvm-project/blob/48bf0a9457fd60d0872d9b9b4804a95c833a72e1/clang/lib/AST/Type.cpp#L3758) with `EST_Unevaluated`.

In the "good" cases I observed, I see that the decl is also loaded with the `noexcept-unevaluated` state, but it gets fixed up later by a [call to `adjustExceptionSpec`](https://github.com/llvm/llvm-project/blob/48bf0a9457fd60d0872d9b9b4804a95c833a72e1/clang/lib/Serialization/ASTReader.cpp#L10656). The update gets [added to `PendingExceptionSpecUpdates` here](https://github.com/llvm/llvm-project/blob/48bf0a9457fd60d0872d9b9b4804a95c833a72e1/clang/lib/Serialization/ASTReaderDecl.cpp#L4749-L4752).

In the "bad" case, the update does not get added to `PendingExceptionSpecUpdates` and hence the call to `adjustExceptionSpec` also does not occur.

# Observations

I made two observations in Clang Discord: [[1](https://discord.com/channels/636084430946959380/636732781086638081/1317290104431185921)] and [[2](https://discord.com/channels/636084430946959380/636732781086638081/1317291980413206608)].

1. [FinishedDeserializating](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L10641) can exit with stuff still in `PendingIncompleteDeclChains`.

`FinishedDeserializing` [calls `finishPendingActions` only if `NumCurrentElementsDeserializing == 1`](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L10647). In [`finishPendingActions`](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L10028), it tries to [clear out `PendingIncompleteDeclChains`](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L10082-L10087). This is done in a loop, which is fine. But, later in `finishPendingActions`, it calls things like `hasBody` [here](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L10296). These can call `getMostRecentDecl` / `getNextRedeclaration` that ultimately calls [`CompleteDeclChain`](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L7719). `CompleteDeclChain` is "called each time we need the most recent declaration of a declaration after the generation count is incremented." according to [this comment](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/AST/ExternalASTSource.h#L213-L215). Anyway, since we're still in `finishPendingActions` with `NumCurrentElementsDeserializing == 1`, calls to `CompleteDeclChain` simply [adds more stuff to `PendingIncompleteDeclChains`](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L7725). ~~I think the emptying out of `PendingIncompleteDeclChains` should actually happen in `FinishedDeserializing`, in [this loop](https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L10657-L10658) instead.~~ (The proposed solution is to do this at the end of `finishPendingActions` instead.

2. `LazyGenerationalUpdatePtr::get` seems a bit dubious. In the `LazyData` case, it does [the following](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/AST/ExternalASTSource.h#L482-L486):

```cpp
      if (LazyVal->LastGeneration != LazyVal->ExternalSource->getGeneration()) {
        LazyVal->LastGeneration = LazyVal->ExternalSource->getGeneration();
        (LazyVal->ExternalSource->*Update)(O);
      }
      return LazyVal->LastValue;
```

so for example, after `markIncomplete`, `LastGeneration` gets set to `0` to force the update. For example, `LazyVal->LastGeneration == 0` and `LazyVal->ExternalSource->getGeneration() == 6`. The `Update` function pointer calls `CompleteDeclChain`, which, if we're in the middle of deserialization, do nothing and just add the decl to `PendingIncompleteDeclChains`. So the update was not completed, but the `LastGeneration` is updated to `6` now... that seems potentially problematic, since subsequent calls will simply return `LazyVal->LastValue` since the generation numbers match now. I would've maybe expected something like:
```
      if (LazyVal->LastGeneration != LazyVal->ExternalSource->getGeneration() &&
          (LazyVal->ExternalSource->*Update)(O)) {
        LazyVal->LastGeneration = LazyVal->ExternalSource->getGeneration();
      }
      return LazyVal->LastValue;
```

where the generation is updated once the intended update actually happens.

# Solution

I verified that in my case, it is indeed the call to `hasBody` inside `finishPendingActions` that bumps the `PendingIncompleteDeclChains` size from `0` to `1`, and also sets the `LazyVal->LastGeneration` to `6` which matches the `LazyVal->ExternalSource->getGeneration()` value of `6`. Later, the iterations over `redecls()` (which calls `getNextRedeclaration`) is expected to trigger the reload, but it **does not** since the generation numbers match.

The proposed solution is to perform the marking of incomplete decl chains at the end of `finishPendingActions`. This way, **all** of the incomplete decls are marked incomplete as a post-condition of `finishPendingActions`. It's also safe to delay this operation since any operation being done within `finishPendingActions` has `NumCurrentElementsDeserializing == 1`, which means that any calls to `CompleteDeclChain` would simply add to the `PendingIncompleteDeclChains` without doing anything anyway.

mpark committed:
[C++20][Modules][Serialization] Delay marking pending incomplete decl chains until the end of `finishPendingActions`.

cor3ntin commented:
Did you try creduce/cvise (although using that with modules might be a bit challenging)
It would be really nice to have a repro so that we have a test case

mpark commented:
> Did you try creduce/cvise (although using that with modules might be a bit challenging) It would be really nice to have a repro so that we have a test case

I did. It did help me reduce it significantly, but it still involves a few modules that are difficult to reduce. I still can make progress by running it on the modules themselves though. I agree it'd be really nice to have a test case.

ChuanqiXu9 commented:
+1. It is best to have more test case. Generally I reduce it by hand in this case.

mpark commented:
> +1. It is best to have more test case. Generally I reduce it by hand in this case.

Yeah, I did already spent quite a bit of effort trying to reduce my hand but have not been successful so far. I'll try again and report back. Would you still review the logic / whether the suggested solution seems sensible?

ChuanqiXu9 commented:
> > +1. It is best to have more test case. Generally I reduce it by hand in this case.
> 
> Yeah, I did already spent quite a bit of effort trying to reduce the creduce result by hand but have not been successful so far 😕. I'll try again and report back. Would you still review the logic / whether the suggested solution seems sensible?

The solution looks neat to me.

mizvekov commented:
Can you reduce your modules test case into a self contained project with build system?

If so, you can use [cvise](https://github.com/marxin/cvise) for reducing a whole project directory recursively in one invocation.

ChuanqiXu9 commented:
> Can you reduce your modules test case into a self contained project with build system?
> 
> If so, you can use [cvise](https://github.com/marxin/cvise) for reducing a whole project directory recursively in one invocation.

I didn't know this before. It sounds great. If you'd like, would you like to write a simple example document to explain how users do in https://clang.llvm.org/docs/StandardCPlusPlusModules.html? I believe it will be pretty helpful.

mpark committed:
[C++20][Modules][Serialization] Add a unit test for #121245.

mpark commented:
Okay folks, I've finally managed to get a reasonable repro as a unit test!

The new unit test, with `Debug` build fails like this:
```
FAIL: Clang :: Modules/pr121245.cpp (8522 of 22154)
******************** TEST 'Clang :: Modules/pr121245.cpp' FAILED ********************
Exit Code: 134

Command Output (stderr):
--
RUN: at line 1: rm -rf /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
+ rm -rf /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
RUN: at line 2: mkdir -p /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
+ mkdir -p /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
RUN: at line 3: split-file /home/mcypark/llvm-project/clang/test/Modules/pr121245.cpp /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
+ split-file /home/mcypark/llvm-project/clang/test/Modules/pr121245.cpp /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
RUN: at line 4: cd /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
+ cd /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp
RUN: at line 6: /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.h   -fcxx-exceptions -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm
+ /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.h -fcxx-exceptions -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm
RUN: at line 9: /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.h   -Wno-experimental-header-units -fcxx-exceptions   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.pcm
+ /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.h -Wno-experimental-header-units -fcxx-exceptions -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.pcm
RUN: at line 13: /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.h   -Wno-experimental-header-units -fcxx-exceptions   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm
+ /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.h -Wno-experimental-header-units -fcxx-exceptions -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm
RUN: at line 17: /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.h   -Wno-experimental-header-units -fcxx-exceptions   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm
+ /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.h -Wno-experimental-header-units -fcxx-exceptions -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm
RUN: at line 21: /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.h   -Wno-experimental-header-units -fcxx-exceptions   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.pcm
+ /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-header-unit -xc++-user-header /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.h -Wno-experimental-header-units -fcxx-exceptions -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm -o /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.pcm
RUN: at line 26: /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-obj /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/main.cpp   -Wno-experimental-header-units -fcxx-exceptions   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.pcm   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm   -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm
+ /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-obj /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/main.cpp -Wno-experimental-header-units -fcxx-exceptions -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm
should not call this with unresolved exception specs
UNREACHABLE executed at /home/mcypark/llvm-project/clang/lib/AST/Type.cpp:3760!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-obj /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/main.cpp -Wno-experimental-header-units -fcxx-exceptions -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm
1.      <eof> parser at end of file
2.      Per-file LLVM IR generation
3.      /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.h:9:3: Generating code for declaration 'EBO<A<long>>::EBO'
 #0 0x0000000005682a18 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/mcypark/llvm-project/llvm/lib/Support/Unix/Signals.inc:798:22
 #1 0x0000000005682e60 PrintStackTraceSignalHandler(void*) /home/mcypark/llvm-project/llvm/lib/Support/Unix/Signals.inc:874:1
 #2 0x0000000005680646 llvm::sys::RunSignalHandlers() /home/mcypark/llvm-project/llvm/lib/Support/Signals.cpp:105:20
 #3 0x00000000056823e1 SignalHandler(int) /home/mcypark/llvm-project/llvm/lib/Support/Unix/Signals.inc:415:1
 #4 0x00007f0f2363e730 __restore_rt (/lib64/libc.so.6+0x3e730)
 #5 0x00007f0f2368bbdc __pthread_kill_implementation (/lib64/libc.so.6+0x8bbdc)
 #6 0x00007f0f2363e686 gsignal (/lib64/libc.so.6+0x3e686)
 #7 0x00007f0f23628833 abort (/lib64/libc.so.6+0x28833)
 #8 0x00000000055bf183 bindingsErrorHandler(void*, char const*, bool) /home/mcypark/llvm-project/llvm/lib/Support/ErrorHandling.cpp:223:55
 #9 0x000000000af32485 clang::FunctionProtoType::canThrow() const /home/mcypark/llvm-project/clang/lib/AST/Type.cpp:3766:12
#10 0x0000000006466ec3 clang::CodeGen::CodeGenFunction::EmitStartEHSpec(clang::Decl const*) /home/mcypark/llvm-project/clang/lib/CodeGen/CGException.cpp:534:32
#11 0x0000000005e53edf clang::CodeGen::CodeGenFunction::StartFunction(clang::GlobalDecl, clang::QualType, llvm::Function*, clang::CodeGen::CGFunctionInfo const&, clang::CodeGen::FunctionArgList const&, clang::SourceLocation, clang::SourceLocation) /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenFunction.cpp:1247:46
#12 0x0000000005e55537 clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenFunction.cpp:1540:16
#13 0x00000000063dd3aa clang::CodeGen::CodeGenModule::codegenCXXStructor(clang::GlobalDecl) /home/mcypark/llvm-project/clang/lib/CodeGen/CGCXX.cpp:212:3
#14 0x000000000601dd57 (anonymous namespace)::ItaniumCXXABI::emitCXXStructor(clang::GlobalDecl) /home/mcypark/llvm-project/clang/lib/CodeGen/ItaniumCXXABI.cpp:4748:46
#15 0x0000000005e820a9 clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:4231:29
#16 0x0000000005e7dd6a clang::CodeGen::CodeGenModule::EmitDeferred() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:3331:31
#17 0x0000000005e7ddb8 clang::CodeGen::CodeGenModule::EmitDeferred() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:3333:7
#18 0x0000000005e7ddb8 clang::CodeGen::CodeGenModule::EmitDeferred() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:3333:7
#19 0x0000000005e7ddb8 clang::CodeGen::CodeGenModule::EmitDeferred() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:3333:7
#20 0x0000000005e7ddb8 clang::CodeGen::CodeGenModule::EmitDeferred() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:3333:7
#21 0x0000000005e7ddb8 clang::CodeGen::CodeGenModule::EmitDeferred() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:3333:7
#22 0x0000000005e70ce7 clang::CodeGen::CodeGenModule::Release() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:853:23
#23 0x0000000006511145 (anonymous namespace)::CodeGeneratorImpl::HandleTranslationUnit(clang::ASTContext&) /home/mcypark/llvm-project/clang/lib/CodeGen/ModuleBuilder.cpp:291:11
#24 0x00000000064fe949 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:242:9
#25 0x00000000093210db clang::ParseAST(clang::Sema&, bool, bool) /home/mcypark/llvm-project/clang/lib/Parse/ParseAST.cpp:191:14
#26 0x000000000681a5e0 clang::ASTFrontendAction::ExecuteAction() /home/mcypark/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1186:11
#27 0x000000000650311d clang::CodeGenAction::ExecuteAction() /home/mcypark/llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:1102:5
#28 0x0000000006819efa clang::FrontendAction::Execute() /home/mcypark/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1076:38
#29 0x000000000673ac6b clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /home/mcypark/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1056:42
#30 0x00000000069cc73e clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /home/mcypark/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:296:38
#31 0x0000000000e3d926 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /home/mcypark/llvm-project/clang/tools/driver/cc1_main.cpp:290:40
#32 0x0000000000e2fa70 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) /home/mcypark/llvm-project/clang/tools/driver/driver.cpp:218:20
#33 0x0000000000e2ff67 clang_main(int, char**, llvm::ToolContext const&) /home/mcypark/llvm-project/clang/tools/driver/driver.cpp:259:26
#34 0x0000000000e67029 main /home/mcypark/llvm-project/build-debug/tools/clang/tools/driver/clang-driver.cpp:17:20
#35 0x00007f0f236295d0 __libc_start_call_main (/lib64/libc.so.6+0x295d0)
#36 0x00007f0f23629680 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x29680)
#37 0x0000000000e2eea5 _start (/home/mcypark/llvm-project/build-debug/bin/clang+0xe2eea5)
/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.script: line 10: 3521235 Aborted                 (core dumped) /home/mcypark/llvm-project/build-debug/bin/clang -cc1 -internal-isystem /home/mcypark/llvm-project/build-debug/lib/clang/21/include -nostdsysteminc -std=c++20 -emit-obj /home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/main.cpp -Wno-experimental-header-units -fcxx-exceptions -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-02.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-05.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-04.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-03.pcm -fmodule-file=/home/mcypark/llvm-project/build-debug/tools/clang/test/Modules/Output/pr121245.cpp.tmp/hu-01.pcm

--

********************
********************
Failed Tests (1):
  Clang :: Modules/pr121245.cpp


Testing Time: 5323.62s

Total Discovered Tests: 46289
  Skipped          :     4 (0.01%)
  Unsupported      :    95 (0.21%)
  Passed           : 46157 (99.71%)
  Expectedly Failed:    32 (0.07%)
  Failed           :     1 (0.00%)
FAILED: tools/clang/test/CMakeFiles/check-clang /home/mcypark/llvm-project/build-debug/tools/clang/test/CMakeFiles/check-clang 
```

cor3ntin commented on clang/lib/Serialization/ASTReader.cpp:
```diff
@@ -10483,6 +10476,13 @@ void ASTReader::finishPendingActions() {
+
+  // For each decl chain that we wanted to complete while deserializing, mark
+  // it as "still needs to be completed".
+  for (unsigned I = 0; I != PendingIncompleteDeclChains.size(); ++I) {
+    markIncompleteDeclChain(PendingIncompleteDeclChains[I]);
+  }
```
```suggestion
  for (Decl* D : PendingIncompleteDeclChains)
    markIncompleteDeclChain(D);
```

I know it's sort of preexisting, but while we are here...

mpark replied:
Yeah, why not

cor3ntin reviewed (commented):
Thanks for the test!

I'll let @ChuanqiXu9 approve.
Presumably, we want to cherry pick that in clang 20?

mpark commented:
> Presumably, we want to cherry pick that in clang 20?

Yeah, I'd think so if it's not too late.

mpark committed:
[C++20][Modules][Serialization] Use range-based `for` instead.

Co-authored-by: cor3ntin <corentinjabot@gmail.com>

dmpolukhin commented on clang/test/Modules/pr121245.cpp:
```diff
@@ -0,0 +1,90 @@
+// RUN: rm -rf %t
```
Please add a comment that the test results in `llvm_unreachable` in debug builds so debug build should be used to investigate problems with the test.

dmpolukhin reviewed (commented):
LGTM but let @ChuanqiXu9 approve.

mizvekov reviewed (approved):
LGTM, Thanks!

mpark commented:
/cherry-pick a9e249f

mpark commented:
/cherry-pick a9e249f64e800fbb20a3b26c0cfb68c1a1aee5e1

zixu-w commented:
Hi @mpark ! Unfortunately with this change we hit a crash in our bootstrap build with modules on macOS:
```
Assertion failed: (D && "missing definition for pattern of instantiated definition"), function hasAcceptableDefinition, file SemaType.cpp, line 9203.
```
Do you mind taking a look? Or temporarily revert this change.
(Compilation command and stack trace in Issue https://github.com/llvm/llvm-project/issues/126973)

mpark commented:
Hi @zixu-w, thanks for the ping. I can't take a look this week, but can on Monday. It would be great if I can look at a small repro though. I'll follow on the issue.

zixu-w commented:
> Hi @zixu-w, thanks for the ping. I can't take a look this week, but can on Monday. It would be great if I can look at a small repro though. I'll follow on the issue.

Thanks! Yes, I'll try to construct a smaller and isolated reproducer.

Is it okay if we revert this change for now in upstream to unblock us?

ChuanqiXu9 commented:
@zixu-w @mpark given the original patch get backported to the release branch and the patch get reverted, I think we should revert the patch in the release branch too.

dmpolukhin commented:
@zixu-w and @ChuanqiXu9 Just want to clarify policies about changes reversion. As far as I understand it was one example in some private codebase with no reproducer publicly available. I can understand and completely agree if it breaks any existing test or llvm-build bot. But in this case I think it would be good to at least wait for the review. Such reversion without providing a reproducer does not allow to fix the issues or even verify that it is related to this PR. This PR fixes another crash with clear reproducers.

ChuanqiXu9 commented:
> @zixu-w and @ChuanqiXu9 Just want to clarify policies about changes reversion. As far as I understand it was one example in some private codebase with no reproducer publicly available. I can understand and completely agree if it breaks any existing test or llvm-build bot. But in this case I think it would be good to at least wait for the review. Such reversion without providing a reproducer does not allow to fix the issues or even verify that it is related to this PR. This PR fixes another crash with clear reproducers.
> 
> Sorry, didn't notice that it is LLVM bootstrap build but still, no still I don't see steps to reproduce in #126973.

CC @AaronBallman for the policy related things.

And my point in the above post is, **if** we revert it in the trunk and it was backported to the release branch, we should revert it in the release branch too. My point is majorly the **if**.

AaronBallman commented:
> > @zixu-w and @ChuanqiXu9 Just want to clarify policies about changes reversion. As far as I understand it was one example in some private codebase with no reproducer publicly available. I can understand and completely agree if it breaks any existing test or llvm-build bot. But in this case I think it would be good to at least wait for the review. Such reversion without providing a reproducer does not allow to fix the issues or even verify that it is related to this PR. This PR fixes another crash with clear reproducers.
> > Sorry, didn't notice that it is LLVM bootstrap build but still, no still I don't see steps to reproduce in #126973.
> 
> CC @AaronBallman for the policy related things.

Our documented policy is at https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy

The promise for a public reproducer is in line with our policy, but reverting a change after it's been in the tree for two weeks is not particularly timely (but not unreasonable either, sometimes it takes a while to churn through downstream team suites). However, it's not clear to me why the revert couldn't have been handled in the downstream given the lack of details on the problem and the fact that upstream does not seem to be obviously broken. For example, the issue could ultimately boil down to interactions with changes carried in the downstream.

That said, I don't think anyone's doing anything wrong (certainly nobody is acting in bad faith), and there was more than a day for reviewers or the patch author to have questioned the revert before the revert was committed.

> Personally I do agree it is frustrating if the patch gets reverted without being able to reproduce it.
> 
> And my point in the above post is, **if** we revert it in the trunk and it was backported to the release branch, we should revert it in the release branch too. My point is majorly the **if**.

I think the rule of thumb should be to revert in the release branch if the revert happens in trunk. CC @tstellar @tru in case they have a different opinion.

zixu-w commented:
Hi @dmpolukhin @ChuanqiXu9 @AaronBallman !

My apologies for the late reversion and not providing more context more promptly. To clarify, the issue does also reproduce in upstream and I've provided reproducing steps in https://github.com/llvm/llvm-project/issues/126973#issuecomment-2657949171. It was only caught in our downstream builds because we don't have the exact configuration to reproduce the problem in upstream builds and CIs.

And our downstream CI had been running into other issues in the past two weeks, which masked the problem and slowed us down in triaging. It also took me a while to understand the assertion and find which commit caused it.

Sorry again for the confusion. I can pick the reversion in the release branch. I'll help follow up with investigating and fixing #126973, and let me know if there's anything else I should do.
